### PR TITLE
esp32: fix build issues after DEV_CFG and DEV_DATA removal

### PR DESCRIPTION
**Pull request by sylvioalves**
PR #41918 added changes that caused regression in ESP32 environment.  
This fixes broken counter driver and removes build warnings in uart and gpio drivers.

### CODE REVIEW

File: drivers/serial/uart_esp32.c [hunk 1]
```diff
@@ -105,7 +105,7 @@ static int uart_esp32_poll_in(const struct device *dev, unsigned char *p_char)
 
 static void uart_esp32_poll_out(const struct device *dev, unsigned char c)
 {
-	const struct uart_esp32_config *config = dev->config;
+	struct uart_esp32_config *config = (struct uart_esp32_config *) dev->config;
```
Comment (gmarull) [inline drivers/serial/uart_esp32.c:108]: Why dropping `const`? Is driver using const data as non-const? If so, please open a bug.

**Review by gmarull — CHANGES_REQUESTED** [review]
Please see my comment

File: drivers/serial/uart_esp32.c (hunk 1 above)
Comment (sylvioalves) [reply drivers/serial/uart_esp32.c:108]: @gmarull All ESP32 `uarl_hal` calls discards cons qualifier. `uart_esp32_config` holds the `hal->dev` driver. That's the reason for removing the const in there. A second option would be moving `hal->dev` instance from `uart_esp32_config` to `uart_esp32_data`, then we keep this as const. Any thoughts?

Comment (gmarull) [reply drivers/serial/uart_esp32.c:108]: If `hal` contains data that is modified at runtime then yes, needs to be moved to `data`.

Comment (sylvioalves) [reply drivers/serial/uart_esp32.c]: @gmarull @glaubermaroto PTAL

**Review by gmarull — APPROVED** [review]

**Review by uLipe — APPROVED** [review]

**Pull request merged**